src/treelearner/cuda/cuda_data_partition.cu

```diff
@@ -0,0 +1,1191 @@
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, false, true, true><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && !missing_is_na && mfb_is_zero && mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, false, true, true><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && !mfb_is_zero && !mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, false, false><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && !mfb_is_zero && !mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, false, false><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && !mfb_is_zero && mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, false, true><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && !mfb_is_zero && mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, false, true><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && mfb_is_zero && !mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, true, false><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && mfb_is_zero && !mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, true, false><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && mfb_is_zero && mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, true, true><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  } else if (missing_is_zero && missing_is_na && mfb_is_zero && mfb_is_na) {
+    const BIN_TYPE* column_data = reinterpret_cast<const BIN_TYPE*>(column_data_pointer);
+    GenDataToLeftBitVectorKernel0<BIN_TYPE, true, true, true, true><<<grid_dim_, block_dim_, 0, cuda_streams_[0]>>>(GenBitVector_ARGS);
+  }
```

Comment:
there are too many conditional templates in this file. It is very easy to make mistakes with too many conditions. 
Is that possible to reduce them by multi-level template or Marcos?

Reply:
Resolved via 4bcaa03